app/classifier/tasks/survey/translations.jsx

```diff
@@ -0,0 +1,66 @@
+import React from 'react';
+import translations from '../../../pages/project/translations';
+
+function SurveyTranslations(props) {
```

Comment:
I find it really hard to follow what this function does 😞 (and I wrote it!)

Reply:
It is really hard. I lost track at the 6th "characteristics". I know I always say this but some destructuring may help?

Reply:
Any specific thoughts on how destructuring would work here? I had the same thought.

Reply:
I've started to rewrite the first block and I've got this
```javascript
const taskStrings = translations.strings.workflow.tasks;
  const { characteristics, choices, questions } = task;
  Object.keys(characteristics).map((characteristicId) => {
    translation.characteristics[characteristicId] = Object.assign({}, characteristics[characteristicId]);
    if (taskStrings[`${props.taskKey}.characteristics.${characteristicId}.label`]) {
      translation.characteristics[characteristicId].label = taskStrings[`${props.taskKey}.characteristics.${characteristicId}.label`];
    }
    translation.characteristics[characteristicId].values = Object.assign({}, characteristics[characteristicId].values);
    Object.keys(task.characteristics[characteristicId].values).map((valueId) => {
      translation.characteristics[characteristicId].values[valueId] = Object.assign({}, characteristics[characteristicId].values[valueId]);
      if (taskStrings[`${props.taskKey}.characteristics.${characteristicId}.values.${valueId}.label`]) {
      translation.characteristics[characteristicId].values[valueId].label =
        taskStrings[`${props.taskKey}.characteristics.${characteristicId}.values.${valueId}.label`];
      }
    });
  });
```
which is slightly less convoluted, but I'd appreciate any advice you can offer to simplify it further.

Reply:
The only thing I can think of is moving the second `.map` block out to its own method. Just to give the eyes a little rest. I'll keep thinking.

Reply:
Roger wrote this for WGE at some point. I think it does what you want.

```
/**
 * Simple is object check.
 * @param item
 * @returns {boolean}
 */
export function isObject(item) {
  return (item && typeof item === 'object' && !Array.isArray(item) && item !== null);
}

/**
 * Deep merge two objects.
 * @param target
 * @param source
 */
export function mergeDeep(target, source) {
  if (isObject(target) && isObject(source)) {
    Object.keys(source).forEach(key => {
      if (isObject(source[key])) {
        if (!target[key]) Object.assign(target, { [key]: {} });
        mergeDeep(target[key], source[key]);
      } else {
        Object.assign(target, { [key]: source[key] });
      }
    });
  }
  return target;
}
```

Reply:
Deep merge will get rid of all the `Object.assign` calls.

I'll have a think about how to reduce the number of maps. Maybe I'm looking at this the wrong way round and the code should be looping through `taskStrings` in order to assign translation strings to the `translation` object.